src/ciphers/macros.rs

```diff
@@ -51,19 +69,44 @@ macro_rules! impl_aead {
                 key: &$crate::ciphers::traits::Key<Self>,
                 nonce: &$crate::ciphers::traits::Nonce<Self>,
                 associated_data: &[u8],
-                tag: &$crate::ciphers::traits::Tag<Self>,
-                ciphertext: &[u8],
                 plaintext: &mut [u8],
+                ciphertext: &[u8],
+                tag: &$crate::ciphers::traits::Tag<Self>,
             ) -> crate::Result<usize> {
                 use aead::{AeadInPlace, NewAead};
 
                 if ciphertext.len() > plaintext.len() {
                     return Err($crate::Error::BufferSize {
+                        name: "plaintext",
                         needs: ciphertext.len(),
                         has: plaintext.len(),
                     });
                 }
 
+                if tag.len() != Self::TAG_LENGTH {
```

Comment:
These length checks shouldn't be necessary when using `GenericArray`? This should be statically asserted.

I think the issue in your example from stronghold is because of this: https://docs.rs/generic-array/0.14.4/src/generic_array/lib.rs.html#172-174

which has the following: `assert_eq!(slice.len(), N::USIZE);`

This is called through the blanket implementation of `TryFrom`: https://doc.rust-lang.org/src/core/convert/mod.rs.html#589

> but the lack of a tag on try_encrypt makes using it inefficient.

Can `try_encrypt` just match the API of `encrypt`?

Reply:
The length checks are necessary because for whatever reason, the assertions don't come through on the length panics in cases where you are doing type conversions.  And as I mentioned in the issue, there are some very strange issues which result from the type conversions where in lengths that should be supported in the `generic_array` type are seemingly not.  

And for the sake of the API, shouldn't there be some internal consistency regarding the error messages?  For instance, right now if I put in the wrong sized tag I get an error from crypto.rs but presuming the assertions are working, if I put in an wrong sized key, ill get that assertion from the panic.  From an aesthetic point of view, it would be nicer to just have a consistent set of descriptive errors. 

I did try matching `try_encrypt` and `encrypt` but there were issues with how the type conversion plays with mutability and returning a reference.  It is possible if we make it so that `try_encrypt` doesn't return the tag however.

Reply:
I completely agree with you on the improvements to the API and I'm all for changing `try_encrypt` so it takes a `&mut tag` argument. 

My concern seeing the length check added is because `GenericArray` is a supporting type for most crypto here and an issue with compile-time lengths not being asserted sounds like a more serious bug?

Reply:
aha

this conversion is no good
```
&mut tag.try_into().unwrap()
```

this one is good
```
(&mut tag).try_into().unwrap()
```

Reply:
Ahh well there you go.  Yeah, I suspect that it was less a bug with the library and more of a rust specific idiosyncrasy.

Reply:
If the `try_*` functions match the other signatures (with your ordering fixes) I think the overall API will be a bit clearer; definitely more consistent. It makes it much easier to move from `encrypt` to `try_encrypt`.

> If you want, I can remove all of these checks?

I would remove the checks - I still don't think they are needed, but maybe add a small comment about this unexpected rust behaviour?

Reply:
Well, ok, Ill remove the checks.  In which case, it makes the most sense to tell people that they should likely use the `try_*` functions as their first priority.